### DISCUSSION TITLE
Documenting Brew Audit Rules

### DISCUSSION THREAD

**Discussion by Justintime50** [Homebrew/discussions#420] [Everyday usage] [answered]
I am building a auto-releasing tool for a personal Homebrew tap and was trying to determine the max-length for the `desc` field as I remember once upon a time that `brew audit` yelled at me for having it too long. Not remembering the error message from months prior, I had to dive into the code and do some searching as after much googling I couldn't find any documented rulesets for the `brew audit` tool.

The question is, can we (should we) document the `brew audit` ruleset? If so, what would be the best approach? I'd love to see some human readable docs as to what to expect as there doesn't appear to be a lot about testing or linting brew formula out there (I very well could just be looking in the wrong places).

Would love to hear opinions on the subject.

**Comment by SMillerDev** [comment, marked as answer]
We're constantly improving the audit rules to make brew work better and I don't think we could keep up with documentation. The best way is just to run the audit to see what it complains about.

**Reply by Justintime50** [reply]
Can you only run audits against live formula? Meaning I can't build a local formula and run audit on that right, it needs to be on GitHub to audit? I had troubles with that in the past.

**Reply by SMillerDev** [reply]
No, I believe `brew audit file.rb` will work. And otherwise you can `brew audit formula` like you would with `brew install` if it's somewhere in the brew directory structure.

**Reply by Rylan12** [reply]
Another reason we probably won't create documentation for it: there are ___a huge number___ of audit/style rules. It's just not worth the effort to document them all when they're constantly evolving.

Personally, I think it's easier to just run `brew audit <formula>` or `brew style <formula>` than it would be to search through each rule in the docs.

If you don't want to create a whole new formula to check a rule, you can always (temporarily) edit an existing formula to check. If I wanted to check the maximum length for `desc`, I would `brew edit youtube-dl` (`youtube-dl` is just my go-to formula for testing for no real reason), add a bunch of characters to the `desc` string, then run `brew audit youtube-dl` or `brew style youtube-dl`. Sure enough, it tells me that the description should be less than 80 characters.

**Reply by Justintime50** [reply]
Fair enough! Thanks for both of your insight!

**Comment by iMichka** [comment]
I worked with 2 strategies in my small career:
- provide a documentation to developers (like the Pep8 text for Python) and a styleguide 
- setup CI with audit rules and tools to enforce the rules

Only solution 2 works. Until someone disables CI when they need to ship their code urgently and break half the production.
